### PR TITLE
[3.3] First integration test for ODCR

### PR DESCRIPTION
**Pull request by dreambeyondorange**
Builds a stack for odcr that includes three reservations and one reservation resource group.  This tests that valid configurations of open ODCR, targeted ODCR, and PG ODCRs can be used

Signed-off-by: Ryan Anderson <ndry@amazon.com>

### Description of changes
* See above

### Tests
* Integration tests ran locally

### References
* https://quip-amazon.com/kaLSAnnbyR7I/Design-Add-Support-for-On-Demand-Capacity-Reservations-ODCRs-in-Cluster-Configuration-File

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/integration-tests/tests/networking/test_on_demand_capacity_reservation.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    odcr_template.add_resource(target_odcr)
+    odcr_template.add_resource(pg_odcr)
+    odcr_template.add_resource(odcr_group)
+
+    odcr_stack = CfnStack(
+        name=generate_stack_name("integ-tests-odcr", request.config.getoption("stackname_suffix")),
+        region=region,
+        template=odcr_template.to_json(),
+    )
+    cfn_stacks_factory.create_stack(odcr_stack)
```
Comment (hanwen-pcluste) [inline tests/integration-tests/tests/networking/test_on_demand_capacity_reservation.py]: Can you refactor the code into a fixture? Using fixture has the benefit of cleaning up resources even if the test fails. The current code does not clean up the ODCR if anything's wrong before Line 157

Comment (dreambeyondorange) [reply tests/integration-tests/tests/networking/test_on_demand_capacity_reservation.py]: Sure, that was something I did initially but pulled back on due to the extra context to get the resource group.  I think it's a bit cleaner now, so I'll do that again!

File: tests/integration-tests/tests/networking/test_on_demand_capacity_reservation.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        cluster, region, "target-odcr-id-pg-cr", odcr_resources["integTestsTargetOdcr"]
+    )
+    _assert_instance_in_capacity_reservation(
+        cluster, region, "target-odcr-arn-pg-cr", odcr_resources["integTestsTargetOdcr"]
+    )
+    _assert_instance_in_capacity_reservation(cluster, region, "pg-odcr-id-cr", odcr_resources["integTestsPgOdcr"])
+    _assert_instance_in_capacity_reservation(cluster, region, "pg-odcr-arn-cr", odcr_resources["integTestsPgOdcr"])
+
+    # need to delete the cluster before deleting placement group
+    cluster.delete()
```
Comment (hanwen-pcluste) [inline tests/integration-tests/tests/networking/test_on_demand_capacity_reservation.py]: This is optional. In the test function definition, you can put `cluster_factory` to the last to make sure it is deleted first
For example:
```
def test_on_demand_capacity_reservation(
    region, pcluster_config_reader, placement_group_stack, odcr_stack, clusters_factory
):
```

Comment (dreambeyondorange) [reply tests/integration-tests/tests/networking/test_on_demand_capacity_reservation.py]: updated ty!

**Review by hanwen-pcluste — APPROVED** [review]
LGTM! Great work

**Pull request merged**